### DISCUSSION TITLE
Use libusb for custom USB HID device

### DISCUSSION THREAD

**Discussion by neuf812v** [libusb/libusb#1472] [Q&A] [answered]
Hi All,
I have HID USB device on STM32 which permanent sends data to PC. I see these data on Windows 10 in USB Monitor Pro, there is bulk transfers. But when I tried to get data with libusb I get nothing, program exit by timeout. Below is my code.

```
int main(int argc, char *argv[]) {

  libusb_context * m_ctx = NULL;
  int r = libusb_init(&m_ctx);
  if (r < 0)
  {
      printf("\nfailed to initialise libusb\n");
      return 1;
  }
  libusb_set_option(m_ctx, LIBUSB_OPTION_LOG_LEVEL, LIBUSB_LOG_LEVEL_DEBUG);


  r = libusb_init_context(/*ctx=*/NULL, /*options=*/NULL, /*num_options=*/0);
  if (r < 0)
  {
      printf("Failed init context\n");
      return r;
  }

  libusb_device_handle* m_hUsb;

  m_hUsb = libusb_open_device_with_vid_pid(m_ctx, 0x1234, 0x5678);
  if (m_hUsb == 0)
  {
    printf("Cannot Open USB Device\n");
    return -1;
  }
  int res;
  res = libusb_kernel_driver_active(m_hUsb, 0);
  if (res == 1)
  {
    res = libusb_detach_kernel_driver(m_hUsb, 0);
    if (res != 0)
    {
      printf("Error detach\n");
      return -1;
    }
    else
    {
        printf("Detached\n");
    }
  }

  libusb_set_configuration(m_hUsb, 1);
  res = libusb_claim_interface(m_hUsb, 0);
  if (res != 0)
  {
     printf("Cannot claim interface - %s\n", libusb_error_name(res));
  }
  uint8_t pData[64];
  int actualLength = 64;
  res = libusb_interrupt_transfer(m_hUsb, 0x81, (uint8_t *)pData, 64, &actualLength, 10000);
  if (res != 0)
  {
          printf("Transfer error - %s\n", libusb_error_name(res));
  }
  else
  {
    for (int i = 0; i < SIZE; i++)
	{
	  printf("%02X ", pData[i]);
	 
	  printf("\n");
  }
}
```

**Comment by mcuee** [comment]
Do you have access to the communication protocol of the device or things like firmware source codes? If not you have to carry out reverse engineering.

**Reply by mcuee** [reply]
BTW, USB HID device does not use Bulk Transfer, only control transfer and interrupt transfer.

**Reply by neuf812v** [reply]
Yes, I have source code to STM32 USB HID device firmware.
Also I tried interrupt transfer but result is the same.

**Reply by mcuee** [reply]
You have to follow your firmware source codes to send the right command to the device. We can not help you much here.

Anyway, if it is a real USB HID device, please use HIDAPI and not libusb.
https://github.com/libusb/libusb/wiki/FAQ#user-content-Does_libusb_support_USB_HID_devices

**Reply by mcuee** [reply, marked as answer]
And you may want to play with hidaptester as well.
https://github.com/todbot/hidapitester

**Reply by mcuee** [reply]
But if you really want to use libusb, you can try to read CH7/CH8 of the USB HID Spec to see if that help.
https://www.usb.org/sites/default/files/hid1_11.pdf

**Reply by neuf812v** [reply]
Thanks, mcuee. I will try. And later I will write about the result.

**Comment by neuf812v** [comment]
I compiled hidapitester in Qt.
Also I corrected CUSTOM_HID_ReportDesc_FS as below:
`

__ALIGN_BEGIN static uint8_t CUSTOM_HID_ReportDesc_FS[USBD_CUSTOM_HID_REPORT_DESC_SIZE] __ALIGN_END =
{	
        0x06, 0x00, 0xff, // Usage Page(Undefined )
	0x09, 0x01, // USAGE (Undefined)
	0xa1, 0x01, // COLLECTION (Application)
	0x15, 0x00, // LOGICAL_MINIMUM (0)
	0x26, 0xff, 0x00, // LOGICAL_MAXIMUM (255)
	0x75, 0x08, // REPORT_SIZE (8)
	0x95, 0x40, // REPORT_COUNT (64)
	0x09, 0x01, // USAGE (Undefined)
	0x81, 0x02, // INPUT (Data,Var,Abs)
	0x95, 0x40, // REPORT_COUNT (64)
	0x09, 0x01, // USAGE (Undefined)
	0x91, 0x02, // OUTPUT (Data,Var,Abs)
	0x95, 0x01, // REPORT_COUNT (1)
	0x09, 0x01, // USAGE (Undefined)
	0xb1, 0x02, // FEATURE (Data,Var,Abs)
	0xC0 /* END_COLLECTION */
};
`
Now my USD HID device read/write data from/to PC in both side.
mcuee, thank you very much again.
